File: algorithms/cpplus/QuickSort.cpp
#include "Driver.h"

template <typename T>
int findPivot(std::vector<T>& input, int begin, int end)
{
  int pivot = input[begin];
  int i = begin - 1;
  int j = end + 1;
  while(true) {
    do ++i; while (input[i] < pivot);
    do --j; while (input[j] > pivot);
    if (i >= j) {
      return j;
    }
    std::swap(input[begin], input[end]);
  }
}

template <typename T>
void quickSort(std::vector<T>& input, int begin, int end)
{
  if (begin < end) {
    T pivot = findPivot(input, begin, end);
    quickSort(input, begin, pivot);
    quickSort(input, pivot + 1, end);
  }
};

void doQuickSort(std::vector<int>& input)
{
  quickSort(input, 0, input.size() - 1);
}


int main(int argn, char* argv[]) 
{
  runMethod(doQuickSort);
}

File: algorithms/java/QuickSort.java
class QuickSort implements SortMethod
{
  @Override
  public void sort(int[] input)
  {
    doSort(input, 0, input.length - 1);
  }

  private int findPivot(int[] input, int begin, int end)
  {
    int pivot = input[begin];
    int i = begin - 1;
    int j = end + 1;
    while(true) {
      do ++i; while (input[i] < pivot);
      do --j; while (input[j] > pivot);
      if (i >= j) {
        return j;
      }
      int temp = input[begin];
      input[begin] = input[end];
      input[end] = temp;
    }
  }

  private void doSort(int[] input, int begin, int end)
  {
    if (begin < end) {
      int pivot = findPivot(input, begin, end);
      doSort(input, begin, pivot);
      doSort(input, pivot + 1, end);
    }
  }

  public static void main(String[] argv)
  {
    Driver.run(new QuickSort());
  }
}
